### PR TITLE
Add groundwork for new versions of federation APIs

### PR DESCRIPTION
**Pull request by erikjohnston**
This will be needed for things like [MSC1794](https://github.com/matrix-org/matrix-doc/pull/1794) (assuming that some form of it gets accepted)

### CODE REVIEW

File: synapse/federation/transport/server.py [hunk 1]
```diff
@@ -286,7 +288,7 @@ def new_func(request, *args, **kwargs):
         return new_func
 
     def register(self, server):
-        pattern = re.compile("^" + PREFIX + self.PATH + "$")
+        pattern = re.compile("^" + self.PREFIX + self.PATH + "$")
```
Comment (richvdh) [inline synapse/federation/transport/server.py]: I'm unconvinced that:

```
    PATH = "/invite/(?P<context>[^/]*)/(?P<event_id>[^/]*)"
    PREFIX = FEDERATION_V2_PREFIX
```

is clearer than

```
    PATH = "/v2/invite/(?P<context>[^/]*)/(?P<event_id>[^/]*)"
```

(in other words: although it's a ballache, would it be better to go through adding `/v1` to the starts of all the `PATH`s?)

File: synapse/federation/transport/client.py [hunk 2]
```diff
@@ -939,20 +939,22 @@ def bulk_get_publicised_groups(self, destination, user_ids):
...
-        prefix (str)
         path (str): String template for the path
         args: ([str]): Args to insert into path. Each arg will be url encoded
 
     Returns:
         str
     """
-    return prefix + path % tuple(urllib.parse.quote(arg, "") for arg in args)
+    return (
+        FEDERATION_V1_PREFIX
```
Comment (richvdh) [inline synapse/federation/transport/client.py]: related to my other comment: we may as well hardcode `/v1` here.

**Comment by erikjohnston** [comment]
I was vaguely copying it from how we do the CS API. Broadly, there the reason to not encode the version in the path is that some endpoints are shared across multiple versions/releases. I don't know if the same will happen to federation API, but it feels somewhat likely.

Though I guess the way that the CS API does this is to actually just have a `PATHS` parameter and use helper functions...

**Comment by erikjohnston** [comment]
In the interest of moving things along, I'm going to merge this as is. I don't think the current way is particularly bad, though I do take your point about whether its worth it. Let's revisit this when we come to figuring out if we want to cut a full `v2/` or `r0/` style API

**Comment by richvdh** [comment]
> Let's revisit this when we come to figuring out if we want to cut a full v2/ or r0/ style API

I feel like I have already figured this out: we do not.

**Pull request merged**